File: day11/solution.ts
import { readFileSync } from "fs";

function findEmptyColumns(universe: string[]){
    const cols: number[] = []
    for(let col = 0; col < universe[0].length; col++){
        let empty = true
        for(let row of universe){
            if(row[col] === '#'){
                empty = false
                break
            }
        }
        if(empty){
            cols.push(col)
        }
    }
    return cols
}
interface galaxy{ x: number; y: number; }
function solve(galaxies: galaxy[], expansionRows: number[], expansionCols: number[], eFactor: number) {
    expansionRows.forEach((row,i) => {
        galaxies = galaxies.map(g => {
            if(g.y > row+i*eFactor){
                return {...g, y: g.y + eFactor}
            }
            return g
        })
    })
    expansionCols.forEach((col,i) => {
        galaxies = galaxies.map(g => {
            if(g.x > col+i*eFactor){
                return {...g, x: g.x+eFactor}
            }
            return g
        })
    })
    let answer = 0
    galaxies.forEach((g1,i) => {
        galaxies.slice(i+1).forEach(g2 => {
            answer += Math.abs(g1.x-g2.x) + Math.abs(g1.y-g2.y)
        })
    })
    return answer
}

const rows = readFileSync(process.argv[2]).toString().split('\n')
const expansionRows = rows.reduce<number[]>((acc,cur,i) => {
    return cur.includes('#') ? acc : [...acc,i]
}, [])
const expansionCols = findEmptyColumns(rows)
const galaxies = rows.flatMap((row,y) => {
    return row.split('')
        .map((v,x) => ({v,x,y}))
        .filter(t => t.v === '#')
})

console.log(`Part 1: ${solve(galaxies, expansionRows, expansionCols, 1)}`)
console.log(`Part 2: ${solve(galaxies, expansionRows, expansionCols, 1000000-1)}`)

File: day11/solution.py

import copy
import sys

def solve(expansion_factor, galaxies, expand_rows, expand_cols):
    for (i,row) in enumerate(expand_rows):
        for g in galaxies:
            if g[1] > (row + i * expansion_factor):
                g[1] += expansion_factor
    for (i,col) in enumerate(expand_cols):
        for g in galaxies:
            if g[0] > col + i * expansion_factor:
                g[0] += expansion_factor
    answer = 0
    for (i, g1) in enumerate(galaxies):
        for g2 in galaxies[i+1:]:
            answer += abs(g1[0]-g2[0]) + abs(g1[1] - g2[1])
    return answer

if __name__ == '__main__':
    with open(sys.argv[1]) as f:
        data = f.read()
    grid = [list(line) for line in data.splitlines()]

    galaxies = []
    for (y,row) in enumerate(grid):
        galaxies.extend([[x,y] for (x,c) in enumerate(row) if c == '#'])

    expand_rows = []
    for i in range(0, len(grid)-1):
        if '#' not in grid[i]:
            expand_rows.append(i)

    expand_cols = []
    grid = list(zip(*grid))
    for i in range(0, len(grid)-1):
        if '#' not in grid[i]:
            expand_cols.append(i)
    
    print(f"Part 1: {solve(1,copy.deepcopy(galaxies),expand_rows,expand_cols)}")
    print(f"Part 2: {solve(1000000-1,copy.deepcopy(galaxies),expand_rows,expand_cols)}")
    